### PR TITLE
[DROOLS-5193] Overall PMML listener design implementation

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @jiripetrlik 

Important notes:
1) Listeners are used only at evaluation time - i.e. involve only PMMLRuntime and model execution
2) every model implementation may - or may not - implement is own listening system
3) it is possible to add listeners on pmmlruntime - in which case they will be available for all pmmlcontext/evaluations, or only inside specific pmmlcontext, in which case they will be available only for that specific context/evaluation
4) I used the term "PMMLStep" because I'm pretty sure I'll have to use PMMLEvent somewhere else, to avoid confusion
5) for the moment being the step information is a simple map - it is not very semantically defined, but at the same time it easily allows model to populate it with model-specific information
6) for the moment being I implemented only the overall execution listening and the miningmodel one; I plan to do other in future PRs
7) it would be interesting to find some feedback/opinion from actual users about what information could be interesting for each model

**Thank you for submitting this pull request**

**NOTE!:** Double check the target branch for this PR.
The default is `main` so it will target Drools 8 / Kogito / Optaplanner 8.
If this PR is not strictly related to drools and kogito project in `drools.git`, it should probably target `7.x`as a branch

**Ports** If a forward-port or a backport is needed, paste the forward port PR here

[link](https://www.example.com)

**JIRA**: 
[link](https://issues.redhat.com/browse/DROOLS-5193)

**referenced Pull Requests**: _(please edit the URLs of referenced pullrequests if they exist)_

* paste the link(s) from GitHub here
* link 2
* link 3 etc.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>

* for a <b>full downstream build</b> 
  * for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  * for <b>github actions</b> job: add the label `run_fdb`

* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

File: kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/runtime/PMMLRuntime.java [hunk 1]
```diff
@@ -48,4 +48,20 @@
      */
     Optional<PMMLModel> getPMMLModel(final String modelName);
 
+    /**
+     * Add the given <code>PMMLListener</code> to the current <code>PMMLRuntime</code>
+     * That listener, in turn, will be added to any <code>PMMLContext</code> passed
+     * to the <code>evaluate</code> method
+     * @param toAdd
+     */
+    void addPMMLListener(final PMMLListener toAdd);
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/runtime/PMMLRuntime.java:57]: What about add a `getListeners` too?

File: kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/enums/PMML_STEP.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Values common to all models, i.e. to overall execution
+ * This should not contain any reference to a specific implementation, to keep modules decoupling
+ */
+public enum PMML_STEP {
+
+    START,
+    AFTER_PREPROCESS,
+    BEFORE_POSTPROCESS,
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/enums/PMML_STEP.java]: Can you find a better name? I don't know if it is clear for the user what this means.
What is PREPROCESS doing? Validation? "Initialization"?

File: kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/models/PMMLStep.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * Interface representing a meaningful <b>step</b> of PMML execution.
+ * The actual meaning will be implemented on a per-model basis
+ */
+public interface PMMLStep extends Serializable {
+
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/models/PMMLStep.java:26]: What about adding a `getId()` method too to identify the step? Maybe an incremental number?

File: kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/test/java/org/kie/pmml/evaluator/core/service/PMMLRuntimeInternalImplTest.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            assertTrue(retrieved.isPresent());
+            commonValuateStep(retrieved.get(), pmml_step, modelMock, requestData);
+        });
+    }
+
+    @Test
+    public void evaluateWithoutListeners() {
+        final PMMLRequestData requestData = getPMMLRequestData();
+        final PMMLContext pmmlContext = new PMMLContextImpl(requestData);
+        pmmlRuntime.evaluate(modelMock, pmmlContext);
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/test/java/org/kie/pmml/evaluator/core/service/PMMLRuntimeInternalImplTest.java]: What are you testing with this test?

File: kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/models/PMMLStep.java (hunk 3 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/models/PMMLStep.java:26]: @danielezonca 
I would prefer to wait for an actual use-case and see what a useful "id" could be. An auto incremental number would be pretty meaningless if global, and cumbersome to manage if execution-specific.

File: kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/runtime/PMMLRuntime.java (hunk 1 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/runtime/PMMLRuntime.java:57]: @danielezonca 
done

File: kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/enums/PMML_STEP.java (hunk 2 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-api/src/main/java/org/kie/pmml/api/enums/PMML_STEP.java]: @danielezonca 
done

File: kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/test/java/org/kie/pmml/evaluator/core/service/PMMLRuntimeInternalImplTest.java (hunk 4 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-evaluator/kie-pmml-evaluator-core/src/test/java/org/kie/pmml/evaluator/core/service/PMMLRuntimeInternalImplTest.java]: @danielezonca 
Removed (leftover)

**Review by danielezonca — APPROVED** [review]

**Pull request merged**